File: typescript/src/registry/skFunctionContextParameterDecorator.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { ISKMethodInfo } from './iSKMethodInfo';
import { ParameterView } from './parameterView';

export interface ISKFunctionContextParameter {
    name: string;
    description?: string;
    defaultValue?: string;
}

/**
 * Decorator to describe the parameters required by a native function.
 * @remarks
 * e.g.
 * @SKFunctionContextParameter({ name: "...", description: "...", defaultValue: "..." })
 */
export function SKFunctionContextParameter(
    parameter: ISKFunctionContextParameter
): (target: any, propertyKey: string, descriptor: PropertyDescriptor) => void {
    return (target: any, propertyKey: string, descriptor: PropertyDescriptor) => {
        if (typeof descriptor.value !== 'function') {
            throw new Error(`@SKFunctionContextParameter(${parameter.name}) decorator can only be applied to methods.`);
        }

        const info = descriptor.value as ISKMethodInfo;
        if (!Array.isArray(info.parameters)) {
            info.parameters = [];
        }
        info.parameters.push(new ParameterView(parameter.name, parameter.description, parameter.defaultValue));
    };
}

File: python/semantic_kernel/skill_definition/sk_function_context_parameter_decorator.py
# Copyright (c) Microsoft. All rights reserved.


def sk_function_context_parameter(
    *, name: str, description: str, default_value: str = ""
):
    """
    Decorator for SK function context parameters.

    Args:
        name -- The name of the context parameter
        description -- The description of the context parameter
        default_value -- The default value of the context parameter
    """

    def decorator(func):
        if not hasattr(func, "__sk_function_context_parameters__"):
            func.__sk_function_context_parameters__ = []

        func.__sk_function_context_parameters__.append(
            {
                "name": name,
                "description": description,
                "default_value": default_value,
            }
        )
        return func

    return decorator
